frame/contracts/src/rent.rs

```diff
@@ -127,75 +177,200 @@ fn try_evict_or_and_pay_rent<T: Trait>(
+		Verdict::Kill => {
+			<ContractInfoOf<T>>::remove(account);
+			child::kill_storage(
+				&alive_contract_info.trie_id,
+				alive_contract_info.child_trie_unique_id(),
+			);
+			<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), false));
+			None
+		}
+		Verdict::Evict { amount } => {
+			if let Some(amount) = amount {
+				amount.withdraw(account);
+			}
 
-		<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), true));
+			// Note: this operation is heavy.
+			let child_storage_root = child::child_root(&alive_contract_info.trie_id);
 
-		return (RentOutcome::Evicted, Some(tombstone_info));
-	}
+			let tombstone = <TombstoneContractInfo<T>>::new(
+				&child_storage_root[..],
+				alive_contract_info.code_hash,
+			);
+			let tombstone_info = ContractInfo::Tombstone(tombstone);
+			<ContractInfoOf<T>>::insert(account, &tombstone_info);
 
-	if pay_rent {
-		let contract_info = ContractInfo::Alive(AliveContractInfo::<T> {
-			rent_allowance: contract.rent_allowance - dues, // rent_allowance is not exceeded
-			deduct_block: current_block_number,
-			..contract
-		});
+			child::kill_storage(
+				&alive_contract_info.trie_id,
+				alive_contract_info.child_trie_unique_id(),
+			);
 
-		<ContractInfoOf<T>>::insert(account, &contract_info);
+			<Module<T>>::deposit_event(RawEvent::Evicted(account.clone(), true));
+			Some(tombstone_info)
+		}
+		Verdict::Charge { amount } => {
+			let contract_info = ContractInfo::Alive(AliveContractInfo::<T> {
+				rent_allowance: alive_contract_info.rent_allowance - amount.peek(),
+				deduct_block: current_block_number,
+				..alive_contract_info
+			});
+			<ContractInfoOf<T>>::insert(account, &contract_info);
 
-		return (RentOutcome::Ok, Some(contract_info));
+			amount.withdraw(account);
+			Some(contract_info)
+		}
 	}
-
-	(RentOutcome::Ok, Some(ContractInfo::Alive(contract)))
 }
 
 /// Make account paying the rent for the current block number
 ///
-/// NOTE: This function acts eagerly.
-pub fn pay_rent<T: Trait>(account: &T::AccountId) -> Option<ContractInfo<T>> {
-	try_evict_or_and_pay_rent::<T>(account, Zero::zero(), true).1
+/// NOTE this function performs eviction eagerly. All changes are read and written directly to
+/// storage.
+pub fn collect_rent<T: Trait>(account: &T::AccountId) -> Option<ContractInfo<T>> {
+	let contract_info = <ContractInfoOf<T>>::get(account);
+	let alive_contract_info = match contract_info {
+		None | Some(ContractInfo::Tombstone(_)) => return contract_info,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+
+	let current_block_number = <frame_system::Module<T>>::block_number();
+	let verdict = consider_case::<T>(
+		account,
+		current_block_number,
+		Zero::zero(),
+		&alive_contract_info,
+	);
+	enact_verdict(account, alive_contract_info, current_block_number, verdict)
 }
 
-/// Evict the account if it should be evicted at the given block number.
+/// Process a snitch that a contract under the given address should be evicted.
```

Comment:
Is `snitch` commonly used in that file? What's the meaning in that context?

Reply:
Ah, I just realized that I meant not "a snitch", but rather a thing that a snitch is doing!

But this is from [this](https://github.com/paritytech/substrate/pull/4733) PR...

Reply:
https://github.com/paritytech/substrate/pull/4790